common/finalizer/finalizer.go

```diff
@@ -118,32 +118,39 @@ func (f *Finalizer) Run(
 	f.mu.Lock()
 	defer f.mu.Unlock()
 
-	totalCount := len(f.callbacks)
+	totalCount := int32(len(f.callbacks))
 	f.logger.Info("finalizer starting",
-		tag.NewInt("items", totalCount),
+		tag.NewInt32("items", totalCount),
 		tag.NewDurationTag("timeout", timeout))
 
-	var completionCounter atomic.Int32
-	var wg sync.WaitGroup
-	wg.Add(totalCount)
-
+	completionChannel := make(chan struct{})
 	ctx, cancel := context.WithTimeout(context.Background(), timeout)
 	defer cancel()
 	go func() {
 		for _, callback := range f.callbacks {
 			// NOTE: Once the pool is stopped due to a timeout, any remaining callbacks will not be invoked anymore.
 			pool.Do(func() {
-				defer wg.Done()
-				defer completionCounter.Add(1)
+				defer func() { completionChannel <- struct{}{} }()
 				_ = callback(ctx)
 			})
 		}
 	}()
 
+	var completionCounter atomic.Int32
 	done := make(chan struct{})
 	go func() {
-		wg.Wait()
```

Comment:
@alexshtin I noticed there was a potential goroutine leak here: if the finalizer times out, this goroutine would have never been completed since the waitgroup would have never been done. So I'm counting the completed items myself now while also listening for the timeout to abandon the goroutine, if needed.